app/models/semester_day.rb

```diff
@@ -0,0 +1,64 @@
+require 'day'
+require 'faculty_semester'
+
+class SemesterDay
+
+  attr_reader :date, :semester
+
+  def initialize(semester, date)
+    @semester = semester.freeze
+    @date = date.freeze
+  end
+
+  def self.resolve(date, faculty_id)
+    if (semester = FacultySemester.find_by_date(date, faculty_id))
+      new semester, date
+    end
+  end
+
+  def wday
+    return date.wday unless period.first_day_override
+
+    @wday ||= begin
+      offset = Day.to_numeric(period.first_day_override) - period.starts_at.wday
+      (date.wday + offset) % 7
+    end
+  end
+
+  def wday_name
+    Day.from_numeric(wday)
+  end
+
+  def week_parity
+    return if period.type != :teaching
+
+    @week_parity ||= begin
+      first_parity = period.first_week_parity == :even ? 0 : 1
+      parity_name(period.starts_at.cweek + date.cweek + first_parity)
```

Comment:
Maybe consider moving `% 2` from `#parity_name` to this line, because it is direct part of this calculation. Hiding part of the calculation somewhere else is not a good idea, especially when the method is called `parity_name`.

Also changing the calculation to `(date.cweek - period.starts_at.cweek) + first_parity` could make things a bit less "magical".